### PR TITLE
Issue 447 v1.2.x fix

### PR DESCRIPTION
**Pull request by dalehenrich**
v1.2.x portion of fix for Issue #447 and Issue #448:
- [x] Blacklist of projects not wanting automatic class initialization is the short term solution (07db1fb)
- [x] #initialize method should never be automatically run in Jadeite, regardless of the black-list setting (824e03f)

### API (see `RwEditToolTest>>testPreferences`)
```smalltalk
"add project to automatic class initialization black list"
Rowan automaticClassInitializationBlackList 
  add: projectName.

"remove project from automatic class initialization black list"
Rowan automaticClassInitializationBlackList 
  remove: projectName
   ifAbsent: [].

"clear automatic class initialization black list"
Rowan clearAutomaticClassInitializationBlackList.

"add project to global automatic class initialization black list"
Rowan glogalAutomaticClassInitializationBlackList 
  add: projectName.

"clear global automatic class initialization black list"
Rowan clearGlobalAutomaticClassInitializationBlackList.

"add project to user automatic class initialization black list"
Rowan userAutomaticClassInitializationBlackList 
  add: projectName.

"clear user automatic class initialization black list"
Rowan clearUserAutomaticClassInitializationBlackList.

"add project to session automatic class initialization black list"
Rowan sessionAutomaticClassInitializationBlackList 
  add: projectName.

"clear session automatic class initialization black list"
Rowan clearSessionAutomaticClassInitializationBlackList.

```

### CODE REVIEW

File: rowan/src/Rowan-Services-Core/RowanClassService.class.st [hunk 1]
```diff
@@ -261,13 +261,16 @@ RowanClassService >> compileMethod: methodString behavior: aBehavior symbolList:
...
+		on: RwExecuteClassInitializeMethodsAfterLoadNotification
+		do: [:ex | ex resume: false ] ] 
+			on: CompileError
+			do: [:ex | ^nil -> (ex gsArguments at: 1)]]
+				on: CompileWarning
+				do: 
+					[:ex | 
+					warnings := ex warningString.
+					ex resume].
 	^[(self compiledMethodAt: method key selector inClass: aBehavior) -> warnings] on: Error
```
Comment (ericwinger) [inline rowan/src/Rowan-Services-Core/RowanClassService.class.st]: ``method`` will become ``false`` and be sent ``key`` which will fail.

**Comment by dalehenrich** [comment]
> As I interpret it, the change wraps rwCompileMethod:category: senders in a new error handler then silently returns false if the error is raised.

> Ideally, the JadeServer class changes won't be called as any method compile should go through services now. So I'm interested in the change to RowanClassService. In this class's change, if the error is raised, caught, and a false returned, the temporary method will be assigned false. Then key will be sent to false and fail.

I believe that you are misinterpreting the action of #resume: ... The **RwExecuteClassInitializeMethodsAfterLoadNotification** handler does not change the behavior of `rwCompileMethod:category:` ... 

The Services tests all passed and I did an interactive test as well (add a self halt to `initialize` method) and things behaved as I expected ...

**Comment by dalehenrich** [comment]
> I don't think that RowanClassService>>compileMethod:behavior: can be sent outside development. Could the new error ever be raised during normal Jadeite usage?

**RwExecuteClassInitializeMethodsAfterLoadNotification** is a notification and has been in use for some time now. The notification is raised during normal Jadeite usage whenever one edits a class-side #initialize method.

In order to avoid sending the #initialize method automatically while doing development in Jadeite all callers of #rwCompileMethod:category: in the Jadeite support code needs to have this change ... 

> Also, is there a reason you didn't do the error handling inside rwCompileMethod:category?

The change in behavior only applies to Jadeite and wrapping the callers in the Jadeite support code is the right place ... The blacklist handling is done within Rowan at a deeper level ...

**Comment by ericwinger** [comment]
Maybe I am misinterpreting but aren't these two code snippets similar? 

The first assigns ``false`` to ``foo`` in the error path. 
``| foo | 
[ foo := 1 zork ] on: Error do:[:ex | ex resume: false].
foo --> false``

won't this assign ``false`` to ``method`` in the error path and get in trouble later?
``[ method := aBehavior rwCompileMethod: methodString category: categorySymbol ]
					on: RwExecuteClassInitializeMethodsAfterLoadNotification
					do: [:ex | ex resume: false ]``  


Note - there may be a bug in my code if ``rwCompileMethod:category:`` can return false. 

I would expect the services tests to pass because the new error isn't raised.

**Comment by dalehenrich** [comment]
> Maybe I am misinterpreting but aren't these two code snippets similar?

No...

> The first assigns false to foo in the error path.
> | foo | [ foo := 1 zork ] on: Error do:[:ex | ex resume: false]. foo --> false

The `resume:` applies the signaller of the **MNU** and the code in the **MNU** handler arranges to return the value ...

> won't this assign false to method in the error path and get in trouble later?
> [ method := aBehavior rwCompileMethod: methodString category: categorySymbol ] on: RwExecuteClassInitializeMethodsAfterLoadNotification do: [:ex | ex resume: false ]

No ... the implementation of the **RwExecuteClassInitializeMethodsAfterLoadNotification** handler uses the value passing into #resume: to decide whether or not to send the #initialize method ... and does not change the behavior of rwCompileMethod:category: ...

> Note - there may be a bug in my code if rwCompileMethod:category: can return false.

Not to worry... #rwCompileMethod:category: does not return `false`

> I would expect the services tests to pass because the new error isn't raised.

This is a good point, I should add test coverage ... You don't appear to have compilation tests in **RowanTestClassServiceTest** have you thought about the pattern that you'd like to use for testing `RowanClassService>>compileMethod:behavior:symbolList:inCategory:`? A passing test should resolve all of your concerns.

**Comment by dalehenrich** [comment]
Speaking of tests, if you take a look at the RwEditToolTest test methods you'll see that several reference **RwExecuteClassInitializeMethodsAfterLoadNotification** and use the `#resume:` pattern and you can look at the return value of the message send within the handler to confirm that the result of that message send is unaffected ...

**Comment by ericwinger** [comment]
>This is a good point, I should add test coverage ... You don't appear to have compilation tests in RowanTestClassServiceTest have you thought about the pattern that you'd like to use for testing RowanClassService>>compileMethod:behavior:symbolList:inCategory:? A passing test should resolve all of your concerns.

The method compilation tests are in the Jadeite client. However, it wouldn't hurt to put some compilation tests in the server. I added issue [https://github.com/GemTalk/Jadeite/issues/356]()

**Comment by dalehenrich** [comment]
@ericwinger, Do you want me to I wait for GemTalk/jadeite#356 before merging this into candidateV1.2?

This issue should be resolved before v1.2.6 is release (in January).

**Comment by ericwinger** [comment]
Go ahead and merge. After you explained that the signaller is making sure we don't return false I don't see a problem. Just need to get some tests written for https://github.com/GemTalk/Jadeite/issues/356 before we release.

**Pull request merged**